cdc/kv/shared_client.go

```diff
@@ -0,0 +1,1233 @@
+	"github.com/pingcap/tiflow/pkg/pdutil"
+	"github.com/pingcap/tiflow/pkg/spanz"
+	"github.com/pingcap/tiflow/pkg/txnutil"
+	"github.com/pingcap/tiflow/pkg/util"
+	"github.com/pingcap/tiflow/pkg/version"
+	"github.com/prometheus/client_golang/prometheus"
+	kvclientv2 "github.com/tikv/client-go/v2/kv"
+	"github.com/tikv/client-go/v2/tikv"
+	pd "github.com/tikv/pd/client"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	grpcerror "google.golang.org/grpc/codes"
+	grpcmeta "google.golang.org/grpc/metadata"
+)
+
+// SubscriptionID comes from `SharedClient.AllocSubscriptionID`.
+type SubscriptionID uint64
+
+// MultiplexingEvent is like model.RegionFeedEvent.
+type MultiplexingEvent struct {
+	model.RegionFeedEvent
+	SubscriptionID SubscriptionID
+	Start          time.Time
+}
+
+// SharedClient is shared in many tables. Methods are thread-safe.
+type SharedClient struct {
+	changefeed model.ChangeFeedID
+	config     *config.KVClientConfig
+	metrics    sharedClientMetrics
+
+	clusterID  uint64
+	filterLoop bool
+
+	pd           pd.Client
+	grpcPool     GrpcPool
+	regionCache  *tikv.RegionCache
+	pdClock      pdutil.Clock
+	lockResolver txnutil.LockResolver
+
+	// requestRangeCh is used to retrieve subscribed table tasks.
+	requestRangeCh *chann.DrainableChann[rangeTask]
+	// regionCh is used to cache region tasks after ranges locked.
+	regionCh *chann.DrainableChann[singleRegionInfo]
+	// regionRouter is used to cache region tasks with rpcCtx attached.
+	regionRouter *chann.DrainableChann[singleRegionInfo]
+	// resolveLockCh is used to retrieve resolve lock tasks.
+	resolveLockCh *chann.DrainableChann[resolveLockTask]
+	errCh         *chann.DrainableChann[regionErrorInfo]
+
+	workers []*sharedRegionWorker
+
+	// only modified in requestRegionToStore so lock is unnecessary.
+	requestedStores map[string]*requestedStore
+
+	totalSpans struct {
+		sync.RWMutex
+		m *spanz.HashMap[SubscriptionID]
+		v map[SubscriptionID]*requestedTable
+	}
+}
+
+type resolveLockTask struct {
+	regionID   uint64
+	maxVersion uint64
+	state      *regionlock.LockedRange
+	enter      time.Time
+}
+
+type rangeTask struct {
+	span           tablepb.Span
+	requestedTable *requestedTable
+}
+
+type requestedStore struct {
+	storeID    uint64
+	storeAddr  string
+	nextStream atomic.Uint32
+	streams    []*requestedStream
+}
+
+type requestedStream struct {
+	streamID uint64
+
+	// To trigger a connect action lazily.
+	preFetchForConnecting *singleRegionInfo
+	requests              *chann.DrainableChann[singleRegionInfo]
+
+	// init is for initializing client and conn asynchronously.
+	init   func(ctx context.Context) error
+	client cdcpb.ChangeData_EventFeedClient
+	conn   *sharedConn
+
+	requestedRegions struct {
+		sync.RWMutex
+		// map[subscriptionID]map[regionID]*regionFeedState
+		m map[SubscriptionID]map[uint64]*regionFeedState
+	}
+}
+
+type requestedTable struct {
+	subscriptionID SubscriptionID
+
+	span      tablepb.Span
```

Comment:
```suggestion
	totalSpan      tablepb.Span
```

Reply:
I prefer `span` because `totalSpan` seems like a collection of table spans.